### ISSUE TITLE
JDK 10+ support

### ISSUE THREAD

**Issue by gutmox** [jwtk/jjwt#399]
I'm using java 9 modules in my project and when I try to use it, I get this compilation error:

```
> Task :auth:compileJava FAILED
error: the unnamed module reads package io.jsonwebtoken.io from both jjwt.api and jjwt.jackson
error: module reactor.core reads package io.jsonwebtoken.io from both jjwt.api and jjwt.jackson
error: module spring.boot.starter reads package io.jsonwebtoken.io from both jjwt.api and jjwt.jackson
error: module jjwt.jackson reads package io.jsonwebtoken.io from both jjwt.api and jjwt.jackson
error: module com.fasterxml.jackson.databind reads package io.jsonwebtoken.io from both jjwt.api and jjwt.jackson
error: module jjwt.impl reads package io.jsonwebtoken.io from both jjwt.api and jjwt.jackson
...
```

My gradle file looks like this:

```
implementation 'io.jsonwebtoken:jjwt-api:0.10.5'
runtime 'io.jsonwebtoken:jjwt-impl:0.10.5','io.jsonwebtoken:jjwt-orgjson:0.10.5'
```

I couldn't make it work or find any workaround to this issue.

**Comment by lhazlewood** [comment]
Wasn't JDK 9 abandoned/deprecated?  They dropped support for it immediately after 9.0.4.  It's not a version JJWT has ever supported and because not even Oracle does anymore, JJWT probably won't either.   JDK 10 and 11 are more appropriate candidates, no?

**Comment by lhazlewood** [comment]
Got it - thanks.  Yeah, we don't build/test on >= JDK 10 just yet, but I'll use this issue to track that work.

**Comment by mcgin** [comment]
The problem here is what is known as a split package problem.  

The package `io.jsonwebtoken.io` is defined in both the jjwt-jackson and jjwt-api modules which is not allowed in J9+.  It can be resolved by moving the two classes in jjwt-jackon to a new package.

`io.jsonwebtoken.io.JacksonDeserializer` -> `io.jsonwebtoken.jackson.JacksonDeserializer`
`io.jsonwebtoken.io.JacksonSerializer` -> `io.jsonwebtoken.jackson.JacksonSerializer`

**Comment by mcgin** [comment]
The same issue applies to the two classes in `jjwt-orgjson`

**Comment by jaapcoomans** [comment]
Bumped into this issue as well. Happy to file a PR with a solution for using JJWT on the module-path. But looking at #400 I want to check the best solution path here first. It's a little bit of a catch 22 because breaking backwards compatibility causes issues for existing users, whereas not solving the issue blocks users wanting to move forward with JPMS.

There are a few options that both solve the split-package issue and keep backwards-compatibility into account:
1. Deprecate and replace the current jackson and orgjson extensions by new ones. That is, not only the classes, but the whole deliverable. As long as jjwt and its extensions are not fully modular (and thus interpreted as an automatic module), only deprecating the offending classes will not help, as their package will still be exported and cause split-package issues.
2. Pull the extensions (back) into the main artifact. That of course has an impact on distribution and transitive dependencies.

Non-options:
1. Go fully modular (and therefore make it possible not to export the offending package). jjwt has some non-modular dependencies. These first need to become proper JPMS modules.
2. Only rename the packages. This will, as mentioned, break backwards compatibility. A little search on github alone taught me that these classes _are_ directly referenced by some projects.

Bottom line I guess that solution number 1 is the most viable option here.

**Comment by lhazlewood** [comment]
@bdemers has this been superseded by another issue that you're working on?  I just want to rectify any duplicate issues if they exist.

**Comment by lhazlewood** [comment]
@bdemers ah, you're addressing this with #488 - my apologies for the noise.
